### PR TITLE
Add basic resharding record migration logic for new cluster (no GC yet)

### PR DESCRIPTION
**Pull request by cy-b**
Basic function for migrating RRs from old clusters to the new cluster and updates config at hash servers. 

Example command for a new cluster to start migration:
`sudo ./dns_server -migrate -hash "172.31.30.97" -config ./new_config.json -port 9121 -cluster "http://172.31.19.166:12379"`

Some functions are not implemented and I will discuss these functions with who picks these tasks:
* Enable/Disable write requests at hash server side.
* Garbage collection of migrated RRs at the old Raft clusters.

### CODE REVIEW

File: dns/dns.go [hunk 1]
```diff
@@ -214,3 +221,216 @@ func serveUDPAPI(store *dnsStore) {
...
+				}
+				// go through Raft to add RRs
+				for _, rrString := range records {
+					if !checkValidRRString(rrString) {
+						// log.Fatal("Received bad RR")
+						errorC <- errors.New("Recevied bad RR")
+						return
+					}
+					// this is async, might cause too much traffic
+					store.ProposeAddRR(rrString)
```
Comment (kevinkassimo) [inline dns/dns.go:381]: Add a small comment to explain why locks are not needed here (due to proposals done through the channel)

File: dns/main.go [hunk 2]
```diff
@@ -29,6 +29,10 @@ func main() {
 	// PLEASE use 9121! hash_server.go assumes this port for forwarding
 	httpAPIPort := flag.Int("port", 9121, "dns HTTP API server port")
 	join := flag.Bool("join", false, "join an existing cluster")
+	migrate := flag.Bool("migrate", false, "need to coordinate migration")
+	// include the port in hash server ip addr
+	hashServer := flag.String("hash", "", "the ip address of a hash server to help with migration")
+	config := flag.String("config", "", "the path to the json file containing the config for this cluster")
```
Comment (kevinkassimo) [inline dns/main.go:35]: Add a comment to explain the format of the config (mostly how it is different from hash_server config)

**Comment by kevinkassimo** [comment]
Also I believe this code is not yet tested? (seems the logic for disabling/reenabling writes is not actually implemented). Be sure to do a minimum test before final merge.

File: dns/dns_store.go [hunk 3]
```diff
@@ -48,6 +49,10 @@ type dnsStore struct {
 	// for sending http Cache requests
 	cluster []string
 	id      int
+	// for adding new cluster and migration
+	config []jsonClusterInfo
```
[the remark this answers is not in the document]
Comment (cy-b) [reply dns/dns_store.go:53]: Yeah that definitely sounds better, no need to change the store state.

File: dns/httpapi.go [hunk 4]
```diff
@@ -119,6 +196,110 @@ func serveHTTPAPI(store *dnsStore, port int, confChangeC chan<- raftpb.ConfChang
...
+					domainNames = append(domainNames, k)
+				}
+			}
+
+			// check if this domain name should be migrated
+			domain := domainNames[index]
+			log.Println("Domain is ", domain)
+			// log.Println(store.lookup.LocateKey([]byte(domain)).String())
+			if shouldMigrate(domain, store) {
+				for _, rrs := range store.store[domainNames[index]] {
```
[the remark this answers is not in the document]
Comment (cy-b) [reply dns/httpapi.go:285]: Are locks needed here? The assumption (implied by the new cluster node's migration logic) is that writes at hash servers are disabled at this point, so all operations should be reads.

File: dns/hash_server/hash_server.go [hunk 5]
```diff
@@ -215,6 +216,69 @@ func serveHashServerHTTPAPI(store *hashServerStore, port int, done chan<- error)
...
+		}
+
+		jsonClusters := make([]jsonClusterInfo, 0)
+		err = json.Unmarshal(body, &jsonClusters)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		// update cluster info in dnsStore
+		store.clusters = intoClusterMap(jsonClusters)
```
Comment (kevinkassimo) [inline dns/hash_server/hash_server.go:264]: Locks here too?

File: dns/httpapi.go [hunk 6]
```diff
@@ -119,6 +144,141 @@ func serveHTTPAPI(store *dnsStore, port int, confChangeC chan<- raftpb.ConfChang
...
+		PrintClusterConfig(store.config)
+		// update consistent
+		cfg := consistent.Config{
+			PartitionCount:    len(store.config),
+			ReplicationFactor: 2, // We are forced to have number larger than 1
+			Load:              3,
+			Hasher:            hasher{},
+		}
+		log.Println("Received cluster update, setting new config")
+		store.lookup = consistent.New(nil, cfg)
```
Comment (kevinkassimo) [inline dns/httpapi.go:203]: Locks ditto.

File: dns/httpapi.go [hunk 7]
```diff
@@ -119,6 +144,141 @@ func serveHTTPAPI(store *dnsStore, port int, confChangeC chan<- raftpb.ConfChang
...
+		for _, clusterJSON := range store.config {
+			store.lookup.Add(clusterToken(clusterJSON.ClusterToken))
```
Comment (kevinkassimo) [inline dns/httpapi.go:205]: Lock ditto

File: dns/httpapi.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (kevinkassimo) [reply dns/httpapi.go:264]: Better still grab the read lock. It is fine to have multiple read locks held anyways.

File: dns/hash_server/hash_server.go (hunk 5 above)
Comment (cy-b) [reply dns/hash_server/hash_server.go:264]: Yes boss

File: dns/httpapi.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (cy-b) [reply dns/httpapi.go:264]: Added read lock around the entire if statement since `shouldMigrate()` also accesses store.

**Pull request merged**